posts/opaleyes-api-breakage-policy.md

```diff
@@ -0,0 +1,385 @@
+
+* *Replacing special purpose functions with general ones:* Removing
+  special purpose functions when more general ones exist (that are
+  still simple) is a good way to keep the API small, reducing user
+  confusion and to easing maintenance burden.
+
+  For example,
+  [`runInsert`](https://hackage.haskell.org/package/opaleye-0.6.0.0/docs/Opaleye-Manipulation.html#v:runInsert)
+  (which could only insert a single element) was replaced with
+  [`runInsertMany`](https://hackage.haskell.org/package/opaleye-0.7.0.0/docs/Opaleye-Manipulation.html#v:runInsertMany)
+  (which could insert a list of elements).  But `runInsertMany` was
+  subsequently found to be insufficiently general (because it did
+  not support `RETURNING` or `ON CONFLICT`) and was replaced by
+  [`runInsert_`](https://hackage.haskell.org/package/opaleye-0.7.0.0/docs/Opaleye-Manipulation.html#v:runInsert_)
+  (which does).  Then, because the name `runInsert` had been freed
+  up, `runInsert_` could be renamed `runInsert` (removing the `_`
+  which was a rough edge in the API).
+
+
+* *Renaming a misnamed function:* To make an API more understandable
+  it can sometimes be useful to choose a better name for a
+  badly-named function.
+
+  For example, there was a function called
+  [`optional`](https://hackage.haskell.org/package/opaleye-0.7.0.0/docs/Opaleye-Table.html#v:optional). Now,
+  `optional` is a very valuable part of the namespace!  Not only is it
+  a common English word, it is [used in
+  `base`](https://www.stackage.org/haddock/lts-18.27/base-4.14.3.0/Control-Applicative.html#v:optional) and
+  it is also used as the name of [a *different*, and more important,
+  Opaleye
+  function](https://hackage.haskell.org/package/opaleye-0.7.0.0/docs/Opaleye-Join.html#v:optional)
+  (with similar functionality to the `base` one)!  Therefore I renamed
+  the former `optional` to `optionalTableField`.
+
+## How to hold the user's hand
+
+Opaleye's *fundamental principle* of API breakage is that the user's
+code should not break when upgrading to Opaleye major version `N+2`
+unless
+
+1. the documentation of (the last point release of) version `N` warned
+of the impending breakage, and
+2. the user received a deprecation warning when compiling against
+`N+1`
+
+Both the documentation warning and the formal deprecation should tell
+the user what she needs to do to repair the breakage.  This way she
+has plenty of advanced notice that she will need to take action, and
+the action to take is clear.  She doesn't risk being thrown into a
+surprise problem solving session without any idea of how much effort
+she will have to spend.
+
+### Deprecation cycle for renaming a function
+
+The fundamental principle suggests, for example, that later versions
+should never change the behaviour of a function.  Instead they should
+add a new function with new behaviour, subsequently deprecate the old
+function, and subsequently remove it.
```

Comment:
Have you considered what is considered "public" behavior? If opaleye changes the SQL it generates, could that be a breaking change? c.f. https://www.hyrumslaw.com/

I think that ideally this would be documented in the project so that users know what might and will not break in a minor release.

For aeson we decided that the textual format of error messages is not part of our API even though changes could  be breaking (test suites are notably often affected). The rationale - or at least, my rationale - is that forcing a new major version to improve error messages would be more costly and i fear it would mean that we simply did not do these improvements.

Reply:
Good question. I agree with your rationale and I don't consider the format of textual things part of the "public" behaviour. In fact the generate SQL is expected to change quite markedly to incorporate opimisations etc., so "public" behaviour is only that which factors through actually running the code on a Postgres instance.